src/validation.cpp

```diff
@@ -322,6 +323,84 @@ static void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfte
 static FILE* OpenUndoFile(const CDiskBlockPos &pos, bool fReadOnly = false);
 
+void MinerLicenses::HandleTx(const CBaseTransaction& tx, const int height) {
+	for (const auto& entry : ExtractLicenseEntries(tx, height)) {
+		if (!FindLicense(entry))
+			AddLicense(entry);
+		else
+			ModifyLicense(entry);
+	}
+}
+
+std::vector<MinerLicenses::LicenseEntry> MinerLicenses::ExtractLicenseEntries(const CBaseTransaction& tx, const int height) {
+	std::vector<MinerLicenses::LicenseEntry> entries;
+
+	for (const auto& vout : tx.vout)
+		if (IsLicenseTxHeader(vout.scriptPubKey))
+			entries.push_back(ExtractLicenseEntry(vout.scriptPubKey, height));
+
+	return entries;
+}
+
+/**
+ * License TX consists of:
+ * OP_RETURN - 1 byte
+ * data size - 1 byte
+ * size of license header - 1 byte
+ * license header - 3 bytes by default
+ * size of script - 1 byte
+ * script - 20-32 bytes
+ * hashrate in PH - 2 bytes
+ */
+MinerLicenses::LicenseEntry MinerLicenses::ExtractLicenseEntry(CScript scriptPubKey, const int height) {
+	const int size = scriptPubKey.size();
+	uint16_t hashRate = scriptPubKey[size - 2] << 8 | scriptPubKey[size - 1];
+	std::string address = "";
+
+	// TODO make it prettier
+	for( int i = 7; i < 7 + scriptPubKey[6] /* size of script */; ++i)
+		address += static_cast<char>(scriptPubKey[i]);
+
+	return MinerLicenses::LicenseEntry{height, hashRate, address};
+}
+
+MinerLicenses::LicenseEntry* MinerLicenses::FindLicense(const MinerLicenses::LicenseEntry& entry) const {
+	auto it = std::find_if(std::begin(licenses), std::end(licenses), [&entry](const MinerLicenses::LicenseEntry& license) {
+		return license.address == entry.address;
+	});
+
+	return it != std::end(licenses) ? const_cast<MinerLicenses::LicenseEntry*>(&*it) : nullptr;
+}
+
+bool MinerLicenses::NeedToUpdateLicense(const MinerLicenses::LicenseEntry& entry) const {
+	auto license = FindLicense(entry);
+	return license != nullptr && license->height < entry.height;
+}
+
+void MinerLicenses::PushLicense(const int height, const uint16_t hashRate, const std::string& address) {
+	auto it = std::find_if(std::begin(licenses), std::end(licenses), [&address](const MinerLicenses::LicenseEntry& obj) {
+		return obj.address == address;
+	});
+
+	if (it == std::end(licenses))
+		licenses.emplace_back(height, hashRate, address);
+}
+
+void MinerLicenses::AddLicense(MinerLicenses::LicenseEntry entry) {
+	if (FindLicense(entry))
+		return;
+
+	licenses.emplace_back(entry.height, entry.hashRate, entry.address);
+}
+
+void MinerLicenses::ModifyLicense(MinerLicenses::LicenseEntry entry) {
+	auto license = FindLicense(entry);
+	if (!license || !NeedToUpdateLicense(entry))
+		return;
+
+	license->hashRate = entry.hashRate;
```

Comment:
I would also add a case for `hashRate==0`, which would result in removing the license from database

Reply:
Done (with new unit test)